server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputerTests.java

```diff
@@ -639,18 +638,45 @@ public void testDesiredBalanceShouldConvergeInABigCluster() {
                 if (primaryNodeId != null) {
                     dataPath.put(new NodeAndShard(primaryNodeId, shardId), "/data");
                     usedDiskSpace.compute(primaryNodeId, (k, v) -> v + thisShardSize);
+                    indexRoutingTableBuilder.addShard(
+                        newShardRouting(
+                            shardId,
+                            primaryNodeId,
+                            null,
+                            true,
+                            STARTED,
+                            AllocationId.newInitializing(inSyncIds.get(shard * (replicas + 1)))
+                        )
+                    );
+                } else {
+                    var lastAllocatedNodeId = randomValueOtherThan(null, () -> randomFrom(remainingNodeIds));
+                    dataPath.put(new NodeAndShard(lastAllocatedNodeId, shardId), "/data");
+                    usedDiskSpace.compute(lastAllocatedNodeId, (k, v) -> v + thisShardSize);
+                    indexRoutingTableBuilder.addShard(
+                        newShardRouting(
+                            shardId,
+                            null,
+                            null,
+                            true,
+                            UNASSIGNED,
+                            RecoverySource.ExistingStoreRecoverySource.INSTANCE,
+                            new UnassignedInfo(
+                                UnassignedInfo.Reason.NODE_LEFT,
+                                null,
+                                null,
+                                0,
+                                0,
+                                0,
+                                false,
+                                UnassignedInfo.AllocationStatus.NO_ATTEMPT,
+                                Set.of(),
+                                lastAllocatedNodeId
+                            ),
+                            AllocationId.newInitializing(inSyncIds.get(shard * (replicas + 1)))
+                        )
+                    );
                 }
 
-                indexRoutingTableBuilder.addShard(
-                    newShardRouting(
-                        shardId,
-                        primaryNodeId,
-                        null,
-                        true,
-                        primaryNodeId == null ? UNASSIGNED : STARTED,
-                        AllocationId.newInitializing(inSyncIds.get(shard * (replicas + 1)))
-                    )
-                );
```

Comment:
This previously generated not empty unassigned shards with either:
* EmptyStoreRecoverySource that does not reduce remaining disk space as it expects the shard is empty
* ExistingStoreRecoverySource that does not reduce remaining disk space as it expects the shard data is already on disk

This is fixed by always relying on ExistingStoreRecoverySource and reserving extra disk space on the node with `lastAllocatedNodeId` for the existing shard data.